\documentclass[11pt]{article}

\usepackage{color}


%: Custom Commands
\newcommand{\note}[1]{{\sf\textcolor{red}{#1}}}

\renewcommand*{\familydefault}{\sfdefault}

%: Title
\title{PCA-seq Package Specification}
\author{JL Kirk}
\date{\today}

\begin{document}
\maketitle

\section{Package Purpose}
The goal of this package is to implement the EIGENSTRAT and PCA-seq method of inferring population structure for both rare and common variants. Specific aims are:
\begin{itemize}
\item Calculate the GRM using EIGENSTRAT or PCA-seq on any subset of the data (rare or common variants)
\item Make use of \texttt{SNPRelate} functions as much as possible for convenience
\item Use \texttt{SNPRelate} syntax as much as possible for user convenience	
\end{itemize}

\subsection{Input Data Types}
\label{input-var}

\begin{itemize}
\item Genotype data: \texttt{SNPGDSFileClass}, SNP GDS file containing all of the relevant genotype data
\item Sample IDs: \texttt{vector}, samples to use in the analysis (NULL implies use all)
\item SNP IDs: \texttt{vector}, SNP ids to use in the analysis (NULL implies use all)
\item MAF filter: \texttt{numeric} or \texttt{vector}, if numeric, assumed to be the minimum MAF value; if vector, assumed to give (min, max) maf values. MAF values are in 0-0.5 \& if two are given, they should not be equal. The first value should be less than the second value.
\item Number of Eigenvectors: \texttt{numeric}, the number of eigenvectors \& values to return. This is should be less than or equal to the number of subjects.
\item Verbose: \texttt{logical}, indicates to show all information printed out from the analysis
\item Remove Monomorphic SNPs: \texttt{logical}, indicates to remove monomorphic SNPs
\item Autosome Only: \texttt{logical} or \texttt{numeric} or \texttt{character}, indicates to use the autosomal chromosomes or the named chromosomes
\item Method: \texttt{string}, indicates which method to use for calculating the GRM
\end{itemize}


\subsection{Output Data Types}
\label{output-var}
The output data type will be an object of the class \texttt{snpgdsPCAClass}, which is a list with the following elements:
\begin{itemize}
\item Sample IDs: \texttt{vector}, the sample ids passed into the function \& used to perform the analysis
\item SNP IDs: \texttt{vector}, the SNP ids passed into the function \& used to perform the analysis
\item Eigenvalues: \texttt{vector}, eigenvalues
\item Eigenvectors: \texttt{matrix}, the eigenvectors in columns
\item Variance Proportion: \texttt{vector}, the proportion of the variance explained by each eigenvector
\item GRM Trace: \texttt{numeric}, the trace of the GRM
\item GRM: \texttt{matrix}, the GRM
\item \note{method: \texttt{string}, the method used to find the GRM}\
\item \note{maf: \texttt{numeric} or \texttt{vector}, the MAF filtering values}
\end{itemize}


\section{Wishlist}
\begin{itemize}
	\item \texttt{seqPCA}: function that performs PCA on sequences (analogous to \texttt{snpgdsPCA} in \texttt{SNPRelate})
	\item \texttt{seqGRM}: function to find the GRM only
	\item \texttt{check.bool}: function to check that the input argument is a \texttt{logical}
	\item \texttt{check.ecnt}: function to check that the \texttt{eigen.cnt} argument is valid
	\item \texttt{check.maf}: function to check that the \texttt{maf} argument is valid
	\item \texttt{check.miss}: function to check that the \texttt{missing.rate} argument is valid
	\item \texttt{check.vect}: function to check if the \texttt{snp.id} and \texttt{samp.id} vectors have length of at least 1 and no duplicates
	\item \texttt{run.grm}: function to compute the appropriate GRM
	\item \texttt{grm.pcaseq}: function to calculate the GRM using PCA-seq
	\item \texttt{grm.eigenstrat}: function to calculate the GRM using EIGENSTRAT
	\item \texttt{get.snps}: function to get the snp ids for one block of snps
	\item \texttt{filter.snps}: function to filter the genotype data based on SNPs
	\item \texttt{filter.monosnps}: function to remove monomorphic SNPs
	\item \texttt{filter.missing}: function to remove SNPs with too much missingness
	\item \texttt{filter.auto}: function to remove sex chromosome SNPs
	\item \texttt{filter.maf}: function to remove SNPs by MAF
	\item \texttt{make.snpgdsPCAClass}: function to construct an object from the \texttt{snpPCAClass}
\end{itemize}


\subsection{Functions from Other Packages}
\begin{itemize}
	\item \texttt{SNPRelate}
	\begin{itemize}
		\item \texttt{print.snpgdsPCA}
		\item \texttt{snpgdsPCAcorr}
	\end{itemize}
	\item \texttt{matrixcalc}
	\begin{itemize}
		\item \texttt{matrix.trace}
	\end{itemize}
	\item \texttt{base}
		\begin{itemize}
			\item \texttt{eigen}
		\end{itemize}
\end{itemize}

\subsection{Function Compatibility with \texttt{SNPRelate}}
\begin{itemize}
	\item \texttt{snpgdsPCALoading}
	\item \texttt{snpgdsPCASampLoading}
\end{itemize}

\section{Function Specification}

\subsection{\texttt{seqPCA}}
This function takes in all of the variables specified in Section \ref{input-var} and returns an object of the class specified in Section \ref{output-var}. Test cases for this function include:
\begin{itemize}
\item A few small cases to check the math
\item Various levels of MAF to check
\item SNP ID filtering
\item Sample ID filtering
\item Monomorphic SNP filtering
\item Autosomal Filtering
\item Try both methods
\end{itemize}

\subsection{\texttt{seqGRM}}
This function takes in all of the variables specified in Section \ref{input-var} and returns sample ids, SNP ids, GRM trace, and the GRM. Test cases for this function include:
\begin{itemize}
\item A few small cases to check the math
\item Various levels of MAF to check
\item SNP ID filtering
\item Sample ID filtering
\item Monomorphic SNP filtering
\item Autosomal Filtering
\item Try both methods
\end{itemize}

\subsection{\texttt{check.bool}}
This function checks if an input parameter is \texttt{logical}. It takes in a \texttt{value} of any class, and returns a \texttt{logical} or an \texttt{error}. Test cases include:
\begin{itemize}
	\item Results for a logical (both TRUE, FALSE)
	\item Results for non-logical
	\item Results for a vector
	\item Results for NA
	\item Results for NULL
	\item Results for NaN
\end{itemize}

\subsection{\texttt{check.ecnt}}
This function checks if an input parameter is a valid number of eigenvalues/eigenvectors. It takes in a \texttt{value} of any class and \texttt{sample.id}, and returns a \texttt{logial} and a \texttt{number} or a\texttt{logical} and an \texttt{error}. If the \texttt{value} is zero, then a warning is returned; if the value is greater than the number of subjects, it is set to the number of subjects and a warning is returned. Test cases include:
\begin{itemize}
	\item Results for correct input
	\item Results for non-integer
	\item Results for negative number
	\item Results for vector
	\item Results for number greater than \# of subjects
	\item Results for 0
	\item Results for NaN
	\item Results for NA
	\item Results for NULL
\end{itemize}


\subsection{\texttt{check.maf}}
This function checks if an input parameter is a valid maf value or range of values. It takes in a \texttt{value} of any class, and returns a \texttt{logial} or an \texttt{error}. Test cases include:
\begin{itemize}
	\item Results for correct input--1 value
	\item Results for correct input--2 values
	\item Results for more than two values
	\item Results for non-numeric values
	\item Results for number outside of $[0, 0.5]$
	\item Results for maf.min $\geq$ maf.max
	\item Results for NaN
	\item Results for NA
	\item Results for NULL
\end{itemize}

\subsection{\texttt{check.miss}}
This function checks if an input parameter is a valid missingness proportion. It takes in a \texttt{value} of any class, and returns a \texttt{logial} or an \texttt{error}. Test cases include:
\begin{itemize}
	\item Results for correct input
	\item Results for more than one value
	\item Results for non-numeric values
	\item Results for number outside of $[0, 1]$
	\item Results for NaN
	\item Results for NA
	\item Results for NULL
\end{itemize}

\subsection{\texttt{run.grm}}
This function calculates the GRM matrix using the appropriate subset of the data and the requested method. It takes in
\begin{itemize}
\item \texttt{gdsobj}: \texttt{SNPGDSFileClass}, a SNP GDS file
\item \texttt{sample.id}: \texttt{vector}, the samples to keep for the analysis
\item \texttt{snp.id}: \texttt{vector}, the SNPs to keep for the anlaysis
\item \texttt{autosome.only}: \texttt{logical}
\item \texttt{remove.monosnp}: \texttt{logical}
\item \texttt{maf}: \texttt{numeric} or \texttt{vector}, if numeric, assumed to be the minimum MAF value; if vector, assumed to give (min, max) maf values. MAF values are in 0-0.5 \& if two are given, they should not be equal
\item \texttt{missing.rate}: \texttt{numeric}
\item \texttt{method}: \texttt{string}
\end{itemize}
This function returns:
\begin{itemize}
	\item grm: \texttt{matrix}
\end{itemize}
Test cases for this function include:
\begin{itemize}
	\item Results for autosome.only is TRUE
	\item Results for autosome.only is FALSE
	\item Results for sample.id removes some samples
	\item Results for snp.id removes some SNPs
	\item Results for remove.monosnp is TRUE
	\item Results for remove.monosnp is FALSE
	\item Results under various MAF
	\item Results for missing.rate is given
	\item Results for both methods
\end{itemize}
\subsection{\texttt{grm.eigenstrat}}
This function calculates the GRM in Pritchard et al 2006. It takes a 
\begin{itemize}
	\item genodat: \texttt{matrix}, the genotype data, SNPs in columns
	\item maf: \texttt{numeric} or \texttt{vector}, the MAF to filter on
	\item snp.id: \texttt{vector}, the snps to filter out
\end{itemize}
This function returns:
\begin{itemize}
	\item grm: \texttt{matrix}
\end{itemize}
\subsection{\texttt{grm.pcaseq}}
This function calculates the GRM using PCA-seq. It takes a 
\begin{itemize}
	\item genodat: \texttt{matrix}, the genotype data, SNPs in columns
	\item maf: \texttt{numeric} or \texttt{vector}, the MAF to filter on
	\item snp.id" \texttt{vector}, the snps to filter out
\end{itemize}
This function returns:
\begin{itemize}
	\item grm: \texttt{matrix}
\end{itemize}

\subsection{\texttt{get.snps}}
This function calculates the indices of the \texttt{i}th block of \texttt{nblock} snps. This function takes:
\begin{itemize}
	\item \texttt{nblock}: \texttt{numeric}, a constant that is the number of SNPs to work with at one time
	\item \texttt{i}: \texttt{numeric}, an integer that indicates which block the snps are in
	\item \texttt{nsnps}: \texttt{numeric}, the total number of snps
\end{itemize}

The function returns a \texttt{vector} of integers of length \texttt{nblock} or less that can be used to subset the \texttt{snp.id} vector.
\subsection{\texttt{filter.snps}}
This function filters the genotype data based on criteria related to the SNPs, after the data has been subset to the SNPs in \texttt{snp.id}. It takes:
\begin{itemize}
	\item \texttt{snp.dat}: a matrix of SNP data
	\item \texttt{autosome.only}: \texttt{logical}
	\item \texttt{remove.monosnp}: \texttt{logical}
	\item \texttt{maf}: \texttt{numeric} or \texttt{vector}, if numeric, assumed to be the minimum MAF value; if vector, assumed to give (min, max) maf values. MAF values are in 0-0.5 \& if two are given, they should not be equal
\item \texttt{missing.rate}: \texttt{numeric}
\end{itemize}
This function returns a \texttt{matrix} of genotype data.
\subsection{\texttt{filter.monosnps}}
This function filters out monomorphic SNPs. It takes:
\begin{itemize}
	\item \texttt{snp.dat}: a matrix of SNP data
\end{itemize}
This function returns a \texttt{matrix} of genotype data.

\subsection{\texttt{filter.missing}}
This function filters out SNPs with too many missing values. It takes:
\begin{itemize}
	\item \texttt{snp.dat}: a matrix of SNP data
	\item \texttt{missing.rate}: \texttt{numeric}
\end{itemize}
This function returns a \texttt{matrix} of genotype data.

\subsection{\texttt{filter.auto}}
This function filters out snps from non-autosomal (sex) chromosomes. It takes:
\begin{itemize}
	\item \texttt{snp.dat}: a matrix of SNP data
\end{itemize}
This function returns a \texttt{matrix} of genotype data.

\subsection{\texttt{filter.maf}}
This function filters the genotype data based on MAF. It takes:
\begin{itemize}
	\item \texttt{snp.dat}: a matrix of SNP data
	\item \texttt{maf}: \texttt{numeric} or \texttt{vector}, if numeric, assumed to be the minimum MAF value; if vector, assumed to give (min, max) maf values. MAF values are in 0-0.5 \& if two are given, they should not be equal
\end{itemize}
This function returns a \texttt{matrix} of genotype data.


\subsection{\texttt{make.snpgdsPCAClass}}
\note{Need to figure out how to calculate the variance proportion.}
This function creates an object of the class \texttt{snpgdsPCACLass}. It takes the inputs:
\begin{itemize}
	\item \texttt{eigen.res}: \texttt{list} of the eigenvalues and eigenvectors
	\item \texttt{grm}: \texttt{matrix}, the GRM
	\item \texttt{sample.id}: \texttt{vector}
	\item \texttt{snp.id}: \texttt{vector}
	\item \texttt{need.genmat}: \texttt{logical}
\end{itemize}
This function returns an object of the class \texttt{snpgdsPCAClass}, which is a list with the fields:
\begin{itemize}
	\item \texttt{sample.id}: \texttt{vector}, a list of sample ids used for the calculation
	\item \texttt{snp.id}: \texttt{vector}, a list of SNPs used for the calculation
	\item \texttt{eigenval}: \texttt{vector}, a list of eigenvalues
	\item \texttt{eigenvect}: \texttt{matrix}, a matrix of eigenvectors
	\item \texttt{varprop}: \texttt{vector}, a vector of the proportion of the variance explained by each eigenvector
	\item \texttt{TraceXTX}: \texttt{numeric}, the trace of the GRM
	\item \texttt{Bayesian}: \texttt{logical}, set to FALSE always
	\item \texttt{genmat}: \texttt{matrix}, the GRM
\end{itemize}
Test cases for this function include:
\begin{itemize}
	\item Correct results for including the GRM
	\item Correct results for not including the GRM
\end{itemize}
\end{document}
